### PR TITLE
fix: Introduce end of stream message to close stream gracefully.

### PR DESCRIPTION
**Pull request by SandPod**
In a cloud environment, streams sometimes seem to drop the last packages of a stream.

A guess is that this is caused by a data race between the subscription cancel call and the `onData` processing.

This fix mitigates the risk of that data race by introducing a last message that is sent through the stream to ensure that the buffer has been completely processed before cancel is called.

The message is passed as an error to circumvent branching in the "hot path" of the stream.

All the awaits have also been removed from the `listen` callback. This ensures that a new message is not processed before all methods have been called, ensuring that the subscription is canceled correctly if required.

Out tests caught this last edge-case by calling an endpoint that threw an exception. Because of the awaits in the `onError` callback, a new message was processed before the subscription could be canceled, ending the stream with `CloseReason.done` instead of `CloseReason.error`.

### Tests

I tried, but failed, to create a test that reproduces the issue observed in the cloud environment. 
To validate the fix, I would like to release it as a 2.3-beta and observe if the issue is resolved.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

None - This should fix an issue but the behavior should stay the same.

### CODE REVIEW

File: packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart [hunk 1]
```diff
@@ -571,21 +578,13 @@ class MethodStreamManager {
...
-          _updateCloseReason(streamKey, CloseReason.done);
-        }
-
-        await subscription.cancel();
-      },
-    );
   }
 }
+
+// Passed as the last message on a stream to indicate that the stream is
```
Comment (christerswahn) [inline packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart:588]: Suggestion: Write as doc comment with 3 /

File: packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart [hunk 2]
```diff
@@ -571,21 +578,13 @@ class MethodStreamManager {
...
 
     outputController.addStream(methodStream).whenComplete(
-      () async {
-        var streamKey = _buildStreamKey(
-          endpoint: methodStreamCallContext.fullEndpointPath,
-          method: methodStreamCallContext.method.name,
-          methodStreamId: methodStreamId,
+          // The stream complete message is sent as an error to circumvent
+          // branching when passing along stream events to the the handler.
+          () => outputController.addError(_StreamComplete()),
```
Comment (christerswahn) [inline packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart:583]: At least on the face of it this appears to play more nice with the underlying stream! :+1:

File: packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart [hunk 3]
```diff
@@ -351,32 +351,39 @@ class MethodStreamManager {
...
 
-        var streamKey = _buildStreamKey(
-          endpoint: methodStreamCallContext.fullEndpointPath,
-          method: methodStreamCallContext.method.name,
-          methodStreamId: methodStreamId,
-        );
         _updateCloseReason(streamKey, CloseReason.error);
 
-        await session.close(error: e, stackTrace: s);
+        session.close(error: e, stackTrace: s);
```
Comment (christerswahn) [inline packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart:378]: Suggestion: Add unawaited to indicate this is intentional, and maybe also add your comment from the PR as code comment here to explain why.

**Review by christerswahn — COMMENTED** [review]
Looking forward to getting to use this! :) Just had a couple of minor comments!

File: packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart (hunk 1 above)
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart:588]: Good catch! will fix!

File: packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart (hunk 3 above)
Comment (SandPod) [reply packages/serverpod/lib/src/server/websocket_request_handlers/helpers/method_stream_manager.dart:378]: Good suggestion! Will fix! 👍

**Review by christerswahn — APPROVED** [review]
LGTM, snyggt!
